# Exploratory Analysis of Climate Data

Exploratory Data Analysis (EDA for short) refers to the process of discovering patterns and relationships within a dataset. It is about making sense of the data at hand and getting to know it before modeling or analysis. It is an important step prior to model building and can help us formulate further questions and areas for investigation. Within this project we are going work with the `training_data.csv` provided by partners of [WiDS Datathon 2023](https://www.kaggle.com/competitions/widsdatathon2023/data).

The goals of EDA are:

- Uncover the data structure and determine how it is coded;
- Inspect and “get to know” the data by summarizing and visualizing it;
- Detect outliers, missing data, and other anomalies and decide how/whether to address these issues;
- Find new avenues for analysis and further research;
- Prepare for model building or analysis, including the following:
    - Check assumptions;
    - Select features;
    - Choose an appropriate method.


As per the publisher WiDS Datathon Overview:

"The WiDS Datathon 2023 focuses on a prediction task involving forecasting sub-seasonal temperatures (temperatures over a two-week period, in our case) within the United States. We are using a pre-prepared dataset consisting of weather and climate information for a number of US locations, for a number of start dates for the two-week observation, as well as the forecasted temperature and precipitation from a number of weather forecast models (we will reveal the source of our dataset after the competition closes). Each row in the corresponds to a single location and a single start date for the two-week period. Your task is to predict the arithmetic mean of the maximum and minimum temperature over the next 14 days, for each location and start date.

You are provided with two datasets:

train_data.csv: the training dataset, where contest-tmp2m-14d__tmp2m, the arithmetic mean of the max and min observed temperature over the next 14 days for each location and start date, is provided test_data.csv: the test dataset, where we withhold the true value of contest-tmp2m-14d__tmp2m for each row.

Target:

contest-tmp2m-14d__tmp2m: the arithmetic mean of the max and min observed temperature over the next 14 days for each location and start date, computed as (measured max temperature + measured mini temperature) / 2".

Let's start by importing some libraries and loading CSV into a DataFrame for inspection and gaining familiarity with the data.

In [1]:
# Basic imports related to analysis
import pandas as pd
import numpy as np

# Imports for visualisation
import seaborn as sns
from matplotlib import pyplot as plt

In [11]:
pd.set_option('display.max_columns', None)
temp_train = pd.read_csv('datasets/temp_train_data.csv')
temp_train.head(5)

index  lat       lon startdate  contest-pevpr-sfc-gauss-14d__pevpr  \
0      0  0.0  0.833333    9/1/14                              237.00   
1      1  0.0  0.833333    9/2/14                              228.90   
2      2  0.0  0.833333    9/3/14                              220.69   
3      3  0.0  0.833333    9/4/14                              225.28   
4      4  0.0  0.833333    9/5/14                              237.24   

   nmme0-tmp2m-34w__cancm30  nmme0-tmp2m-34w__cancm40  \
0                     29.02                     31.64   
1                     29.02                     31.64   
2                     29.02                     31.64   
3                     29.02                     31.64   
4                     29.02                     31.64   

   nmme0-tmp2m-34w__ccsm30  nmme0-tmp2m-34w__ccsm40  nmme0-tmp2m-34w__cfsv20  \
0                    29.57                    30.73                    29.71   
1                    29.57                    30.73                    29.71   
2                    29.57                    30.73                    29.71   
3                    29.57                    30.73                    29.71   
4                    29.57                    30.73                    29.71   

   nmme0-tmp2m-34w__gfdlflora0  nmme0-tmp2m-34w__gfdlflorb0  \
0                        31.52                        31.68   
1                        31.52                        31.68   
2                        31.52                        31.68   
3                        31.52                        31.68   
4                        31.52                        31.68   

   nmme0-tmp2m-34w__gfdl0  nmme0-tmp2m-34w__nasa0  nmme0-tmp2m-34w__nmme0mean  \
0                   30.56                   29.66                       30.46   
1                   30.56                   29.66                       30.46   
2                   30.56                   29.66                       30.46   
3                   30.56                   29.66                       30.46   
4                   30.56                   29.66                       30.46   

   contest-wind-h10-14d__wind-hgt-10  nmme-tmp2m-56w__cancm3  \
0                           31246.63                    28.3   
1                           31244.78                    28.3   
2                           31239.27                    28.3   
3                           31232.86                    28.3   
4                           31226.16                    28.3   

   nmme-tmp2m-56w__cancm4  nmme-tmp2m-56w__ccsm3  nmme-tmp2m-56w__ccsm4  \
0                   29.47                  27.13                  27.36   
1                   29.47                  27.13                  27.36   
2                   29.47                  27.13                  27.36   
3                   29.47                  27.13                  27.36   
4                   29.47                  27.13                  27.36   

   nmme-tmp2m-56w__cfsv2  nmme-tmp2m-56w__gfdl  nmme-tmp2m-56w__gfdlflora  \
0                  27.71                 28.25                       27.7   
1                  27.71                 28.25                       27.7   
2                  27.71                 28.25                       27.7   
3                  27.71                 28.25                       27.7   
4                  27.71                 28.25                       27.7   

   nmme-tmp2m-56w__gfdlflorb  nmme-tmp2m-56w__nasa  nmme-tmp2m-56w__nmmemean  \
0                      28.72                 28.38                     28.11   
1                      28.72                 28.38                     28.11   
2                      28.72                 28.38                     28.11   
3                      28.72                 28.38                     28.11   
4                      28.72                 28.38                     28.11   

   contest-rhum-sig995-14d__rhum  nmme-prate-34w__cancm3  \
0                          81.72  

In [12]:
temp_train.tail(5)

index  lat       lon startdate  contest-pevpr-sfc-gauss-14d__pevpr  \
375729  375729  1.0  0.866667   8/27/16                              312.05   
375730  375730  1.0  0.866667   8/28/16                              305.82   
375731  375731  1.0  0.866667   8/29/16                              311.62   
375732  375732  1.0  0.866667   8/30/16                              304.54   
375733  375733  1.0  0.866667   8/31/16                              295.29   

        nmme0-tmp2m-34w__cancm30  nmme0-tmp2m-34w__cancm40  \
375729                     23.13                      27.2   
375730                     23.13                      27.2   
375731                     23.13                      27.2   
375732                     23.13                      27.2   
375733                     23.13                      27.2   

        nmme0-tmp2m-34w__ccsm30  nmme0-tmp2m-34w__ccsm40  \
375729                    20.25                    24.43   
375730                    20.25                    24.43   
375731                    20.25                    24.43   
375732                    20.25                    24.43   
375733                    20.25                    24.43   

        nmme0-tmp2m-34w__cfsv20  nmme0-tmp2m-34w__gfdlflora0  \
375729                    18.35                        23.35   
375730                    18.35                        23.35   
375731                    18.35                        23.35   
375732                    18.35                        23.35   
375733                    18.35                        23.35   

        nmme0-tmp2m-34w__gfdlflorb0  nmme0-tmp2m-34w__gfdl0  \
375729                        23.59                   19.43   
375730                        23.59                   19.43   
375731                        23.59                   19.43   
375732                        23.59                   19.43   
375733                        23.59                   19.43   

        nmme0-tmp2m-34w__nasa0  nmme0-tmp2m-34w__nmme0mean  \
375729                   23.23                       22.55   
375730                   23.23                       22.55   
375731                   23.23                       22.55   
375732                   23.23                       22.55   
375733                   23.23                       22.55   

        contest-wind-h10-14d__wind-hgt-10  nmme-tmp2m-56w__cancm3  \
375729                           31484.23                   18.55   
375730                           31476.67                   17.92   
375731                           31464.71                   17.29   
375732                           31448.83                   16.02   
375733                           31434.76                   15.39   

        nmme-tmp2m-56w__cancm4  nmme-tmp2m-56w__ccsm3  nmme-tmp2m-56w__ccsm4  \
375729                   23.01                  17.32                  22.03   
375730                   22.35                  16.89                  21.49   
375731                   21.70                  16.45                  20.94   
375732                   20.39                  15.57                  19.85   
375733                   19.74                  15.14                  19.30   

        nmme-tmp2m-56w__cfsv2  nmme-tmp2m-56w__gfdl  \
375729                  14.94                 18.95   
375730                  14.56                 18.50   
375731                  14.17                 18.05   
375732                  13.40                 17.15   
375733                  13.02                 16.70   

        nmme-tmp2m-56w__gfdlflora  nmme-tmp2m-56w__gfdlflorb  \
375729                      17.44                      18.35   
375730                      16.86                      17.76   
375731                      16.27                      17.17   
375732                      15.11                      15.99   
375733                      14.53                      15.40   

        nmme-tmp2m-56w__nasa  nmme-tmp2m-56w__nmmem

In [15]:
temp_train.shape

(375734, 246)

It is a fairly big dataset with `246` columns and `375734` rows. Let's get a short technical overview of the dataset.

In [3]:
temp_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375734 entries, 0 to 375733
Columns: 246 entries, index to wind-vwnd-925-2010-20
dtypes: float64(240), int64(4), object(2)
memory usage: 705.2+ MB


Result: 
- The majority of the values have data type `float` which is a floating point number. We can notice that some of them are temperature data, some - predictions from other models and the rest are presumably wind speed values. 
- 2 of the columns contain data with the type `object`. One of them is visible from the start, it is a `startdate` and it contains a measurement beginning date. The other one we can find out using the following function:

In [31]:
def check_dtype(dtype, num_values):
    l = []
    for x in temp_train.columns:
        if temp_train[x].dtype == dtype:
            l.append(temp_train[x][:num_values])
    return l
        
check_dtype('object',3)

[0    9/1/14
 1    9/2/14
 2    9/3/14
 Name: startdate, dtype: object,
 0    BSh
 1    BSh
 2    BSh
 Name: climateregions__climateregion, dtype: object]

Result: 
Besides `startdate`, the `object` data type has column `climateregions__climateregion`. Let's check short statistic for this value.

In [29]:
temp_train.climateregions__climateregion.describe()

count     375734
unique        15
top          BSk
freq      139621
Name: climateregions__climateregion, dtype: object

Result: 
We can see that there are apparently no `NaN`s within this column, there are 15 unique values and the most frequently used (139621 times) is `BSk`. 

How many times each value comes across within the dataset? 

In [21]:
temp_train.climateregions__climateregion.value_counts()

BSk    139621
Dfb     52632
Cfa     51901
Csb     40936
Dfa     22661
BWk     13889
Dfc     12427
BWh      9503
Csa      9503
Dsb      8041
BSh      5117
Cfb      4386
Dsc      2924
Dwa      1462
Dwb       731
Name: climateregions__climateregion, dtype: int64

Let's proceed to the columns with the `int64` data type. There were 4 of them, let's check how they look and what are they.

In [30]:
check_dtype('int64', 3)

[0    0
 1    1
 2    2
 Name: index, dtype: int64,
 0    4
 1    4
 2    4
 Name: mjo1d__phase, dtype: int64,
 0    56
 1    56
 2    56
 Name: mei__meirank, dtype: int64,
 0    4
 1    4
 2    4
 Name: mei__nip, dtype: int64]

In [6]:
temp_train.describe()

index            lat            lon  \
count  375734.000000  375734.000000  375734.000000   
mean   187866.500000       0.592766       0.517964   
std    108465.207359       0.251744       0.272059   
min         0.000000       0.000000       0.000000   
25%     93933.250000       0.409091       0.300000   
50%    187866.500000       0.590909       0.533333   
75%    281799.750000       0.818182       0.766667   
max    375733.000000       1.000000       1.000000   

       contest-pevpr-sfc-gauss-14d__pevpr  nmme0-tmp2m-34w__cancm30  \
count                       375734.000000             375734.000000   
mean                           276.744064                 10.801682   
std                            198.085182                 10.952417   
min                             -2.310000                -14.960000   
25%                            104.062500                  1.550000   
50%                            250.320000                  9.950000   
75%                            420.010000                 20.230000   
max                           1185.140000                 36.080000   

       nmme0-tmp2m-34w__cancm40  nmme0-tmp2m-34w__ccsm30  \
count             375734.000000            359800.000000   
mean                  12.726739                10.099323   
std                   11.073553                10.339626   
min                  -14.970000               -11.720000   
25%                    3.650000                 1.640000   
50%                   11.890000                 9.490000   
75%                   22.110000                18.740000   
max                   36.030000                33.390000   

       nmme0-tmp2m-34w__ccsm40  nmme0-tmp2m-34w__cfsv20  \
count            375734.000000            375734.000000   
mean                 11.638229                10.733886   
std                  10.574276                 9.403434   
min                 -17.560000               -13.720000   
25%                   3.080000                 3.390000   
50%                  11.600000                10.890000   
75%                  20.360000                18.330000   
max                  34.640000                35.750000   

       nmme0-tmp2m-34w__gfdlflora0  nmme0-tmp2m-34w__gfdlflorb0  \
count                375734.000000                375734.000000   
mean                     11.772646                    11.694567   
std                      10.689967                    10.681607   
min                     -13.330000                   -13.520000   
25%                       2.970000                     2.910000   
50%                      11.280000                    11.240000   
75%                      20.300000                    20.160000   
max                      38.350000                    38.450000   

       nmme0-tmp2m-34w__gfdl0  nmme0-tmp2m-34w__nasa0  \
count           375734.000000           375734.000000   
mean                11.037192               12.226311   
std                 10.649950               11.744265   
min                -11.720000              -22.510000   
25%                  2.160000                3.110000   
50%                 10.100000               13.240000   
75%                 19.770000               22.500000   
max                 36.120000               38.220000   

       nmme0-tmp2m-34w__nmme0mean  contest-wind-h10-14d__wind-hgt-10  \
count               375734.000000                      375734.000000   
mean                    11.483127                       31071.557568   
std                     10.580067                         343.019882   
min                    -14.680000                       29943.150000   
25%                      2.690000                       30811.990000   
50%                     11.200000                       31016.820000   
75%                     20.420000                       31374.780000   
max                     34.880000                       31803.030000   

       nmme-tmp2m-56w__cancm3  nmme-tmp2m-56w__c

In [7]:
def with_nans(dataset):
    
    with_nan = dataset.isna().sum()
    
    if with_nan[with_nan > 0].empty:
        print('No NaN values found.')
    else:
        return with_nan[with_nan > 0] 

with_nans(temp_train)

nmme0-tmp2m-34w__ccsm30    15934
nmme-tmp2m-56w__ccsm3      10280
nmme-prate-34w__ccsm3       8738
nmme0-prate-56w__ccsm30    15934
nmme0-prate-34w__ccsm30    15934
nmme-prate-56w__ccsm3      10280
nmme-tmp2m-34w__ccsm3       8738
ccsm30                     15934
dtype: int64

In [8]:
def average_nans(dataset):
    # Applying only on variables with NaN values
    for i in dataset.columns[dataset.isnull().any(axis=0)]:     
        dataset[i].fillna(dataset[i].mean(),inplace=True)
    return dataset

temp_train = average_nans(temp_train)

In [9]:
with_nans(temp_train)

No NaN values found.


TBC.